### ISSUE TITLE
Follow proper OAuth2

### ISSUE THREAD

**Issue by balloob** [tchellomello/python-ring-doorbell#144]
I am in contact with Ring to get Home Assistant their own client ID / secret for account linking. We will make this available via the new account linking service. 

But for this to be done, the authentication part needs to be extracted from the ring object so that it can hook into the account linking service. Authentication part is the refreshing tokens and making authenticated requests part. It does not cover the obtaining token code.

Home Assistant has documentation on how this lib can be structured: https://developers.home-assistant.io/docs/en/api_lib_auth.html#sync-example-1 . You can also choose to just put the `query` function into a new auth class, but it means you need to do a bit more work integrating it into an OAuth2 config flow in Home Assistant.

**Comment by balloob** [comment]
I have been in touch with Ring. The reason they are actively blocking this lib right now is that this library is not following the proper OAuth2 authentication. Whenever the token expires, we should get a new access token using the refresh token that was returned, instead of fetching a new access/refresh token using username/password. 

The linked tutorial above describes exactly how this can be added to this lib.

Ring wants to work with us, but we need to follow the spec. Until we do, they will keep blocking us.

**Comment by arsaboo** [comment]
Indeed, the last fix was blocked again today.

**Comment by tekmaven** [comment]
@balloob could you ask Ring to give the project an extension to implement this?  It is the holiday time and many people's smart homes are breaking who have invested in the Ring ecosystem.

**Comment by isabellaalstrom** [comment]
As I have understood it there's already a way to implement this. Someone just have to make amendments to the lib.

**Comment by Scope666** [comment]
The workaround (for now) is here:

https://github.com/tchellomello/python-ring-doorbell/issues/146#issuecomment-568664517

**Comment by asantaga** [comment]
This oAuth issue is just for Home Assistant right? or are they asking we do this for the base API too?

**Comment by steve-gombos** [comment]
So does Ring want us to use the client id/secret now to establish connection to their token service or is implementing the refresh token after initial authentication what they want?  

I've forked this a few days ago and started working on adding the usage of the refresh token to generate a new access token.  Specifically was working on this for the 2FA stuff since it isn't supported right now, but seems to all be getting lumped together with this issue now.  

Link to my fork if anyone cares: https://github.com/steve-gombos/python-ring-doorbell

I'm pretty new to Python and coding against Home Assistant, so trying to start somewhere.

**Comment by xernaj** [comment]
> So does Ring want us to use the client id/secret now to establish connection to their token service or is implementing the refresh token after initial authentication what they want?
> 
> I've forked this a few days ago and started working on adding the usage of the refresh token to generate a new access token. Specifically was working on this for the 2FA stuff since it isn't supported right now, but seems to all be getting lumped together with this issue now.
> 
> Link to my fork if anyone cares: https://github.com/steve-gombos/python-ring-doorbell
> 
> I'm pretty new to Python and coding against Home Assistant, so trying to start somewhere.

It sounds like they want proper oauth 2 with refresh token and account linking via some service which will generate a client id/secret for each user. This way, your username/password will not be needed after the first time. I'm assuming this will be a lot more secure than the username/password combo or will at least look more secure due to media scrutiny, etc..

**Comment by steve-gombos** [comment]
> > So does Ring want us to use the client id/secret now to establish connection to their token service or is implementing the refresh token after initial authentication what they want?
> > I've forked this a few days ago and started working on adding the usage of the refresh token to generate a new access token. Specifically was working on this for the 2FA stuff since it isn't supported right now, but seems to all be getting lumped together with this issue now.
> > Link to my fork if anyone cares: https://github.com/steve-gombos/python-ring-doorbell
> > I'm pretty new to Python and coding against Home Assistant, so trying to start somewhere.
> 
> It sounds like they want proper oauth 2 with refresh token and account linking via some service which will generate a client id/secret for each user. This way, your username/password will not be needed after the first time. I'm assuming this will be a lot more secure than the username/password combo or will at least look more secure due to media scrutiny, etc..

Well the forked code I updated does this and also adds 2FA, but it could probably use some clean up. This repo doesn't use the oauth lib like suggested in this issue, and I'm not sure how well it will play with the Ring 2FA., I'd be willing take a stab at that when I have some time here in a few days.  Assuming we just need to setup an auth class that HA can implement with correctly.

**Comment by balloob** [comment]
@steve-gombos haven't looked at your code, but based on you comments about oauth2 lib just a quick comment. 

There is a split in auth responsibility. 

There is a piece of code that you give username/password/2FA and it gives you a refresh/access token pair. This is an independent piece of code.

The API can now be initialized with just the refresh/access token using `requests_oauthlib`, and it will refresh the access token automatically once it expires and allow easy linking with HA.

**Comment by tchellomello** [comment]
Hello, applying the workaround proposed on #149 did help to temporarily address the issue, however, it will matter of time to get it blocked again IMHO.

Reviewing the code this night I actually found a pretty nasty bug (sorry about that) https://github.com/tchellomello/python-ring-doorbell/blob/master/ring_doorbell/__init__.py#L90-L104 which causes the code to generate a new Bearer token for every single request. No wonder why this is getting blocked.

I think the proper resolution will be to store the access_token and the refresh_token as an attribute and then create a mechanism or use the oauth2 library to control and refresh the token. We could create an independent class to treat this or just embed it into the code.

I'll try to read more about how to properly use the library. Any help is very welcome too!

**Comment by steve-gombos** [comment]
@tchellomello I did notice this when I forked, which is part of the fixes I implemented in my fork. It's pretty rough but seems to work. Basically would save that access token and refresh token into the pickle file (not sure if this is the right place to). It would then load those from the pickle file any time and use the refresh token grant type vs the password grant type.  

At the same time I have also been reviewing how to implement the oauth2 library. I have some auth code setup locally, but I'm not sure the best way to integrate that into the existing Ring class. Either pass in the Auth object? or just the token?  I could share what I have here if needed.
